File: java/src/test/java/com/kcl/ExecProgramFormatTest.java
package com.kcl;

import com.kcl.api.API;
import com.kcl.api.Spec.ExecProgramArgs;
import com.kcl.api.Spec.ExecProgramResult;

import org.junit.Assert;
import org.junit.Assume;
import org.junit.Test;

/**
 * Tests for the source-map fields introduced alongside
 * kcl-lang/kcl#1546. Combines:
 *   1. Pure protobuf round-trip — verifies the regenerated Spec.java
 *      encodes / decodes the new fields without touching the runtime.
 *   2. End-to-end execution — runs the bundled schema.k through
 *      API().execProgram with format="json" and sourcemap_output set,
 *      asserting on the actual runtime output.
 *
 * Covers:
 *   - ExecProgramArgs.format          (field 20)
 *   - ExecProgramArgs.error_format    (field 19, pre-existing sanity check)
 *   - ExecProgramArgs.sourcemap_output (field 22)
 *   - ExecProgramResult.sourcemap     (field 5)
 */
public class ExecProgramFormatTest {

    private static final String TEST_FILE = "./src/test_data/schema.k";

    @Test
    public void testExecProgramArgsFormatRoundTrip() throws Exception {
        ExecProgramArgs args = ExecProgramArgs.newBuilder()
                .setFormat("json")
                .setErrorFormat("sarif")
                .setSourcemapOutput("/tmp/out.js.map")
                .build();

        ExecProgramArgs decoded = ExecProgramArgs.parseFrom(args.toByteString());

        Assert.assertEquals("json", decoded.getFormat());
        Assert.assertEquals("sarif", decoded.getErrorFormat());
        Assert.assertTrue("sourcemap_output presence", decoded.hasSourcemapOutput());
        Assert.assertEquals("/tmp/out.js.map", decoded.getSourcemapOutput());
    }

    @Test
    public void testExecProgramResultSourcemapRoundTrip() throws Exception {
        ExecProgramResult result = ExecProgramResult.newBuilder()
                .setJsonResult("{\"a\": 1}")
                .setYamlResult("a: 1")
                .setSourcemap("{\"version\":3,\"sources\":[]}")
                .build();

        ExecProgramResult decoded = ExecProgramResult.parseFrom(result.toByteString());

        Assert.assertEquals("{\"a\": 1}", decoded.getJsonResult());
        Assert.assertEquals("a: 1", decoded.getYamlResult());
        Assert.assertTrue("sourcemap presence", decoded.hasSourcemap());
        Assert.assertEquals("{\"version\":3,\"sources\":[]}", decoded.getSourcemap());
    }

    // End-to-end: format=json must populate only json_result.
    @Test
    public void testExecProgramFormatJson() throws Exception {
        ExecProgramArgs args = ExecProgramArgs.newBuilder()
                .addKFilenameList(TEST_FILE)
                .setFormat("json")
                .build();
        ExecProgramResult result = new API().execProgram(args);

        Assert.assertFalse(
                "format=json: json_result must be populated",
                result.getJsonResult().isEmpty());
        Assert.assertEquals(
                "format=json: yaml_result must be empty",
                "", result.getYamlResult());
        Assert.assertTrue(
                "format=json: json_result must contain replicas key",
                result.getJsonResult().contains("replicas"));
    }

    // End-to-end: sourcemap_output must populate result.sourcemap. Older
    // runtimes that don't yet emit source maps skip via Assume so this
    // test doesn't fail under a stale kcl-api.
    @Test
    public void testExecProgramSourcemapOutput() throws Exception {
        ExecProgramArgs args = ExecProgramArgs.newBuilder()
                .addKFilenameList(TEST_FILE)
                .setSourcemapOutput("/tmp/out.js.map")
                .build();
        ExecProgramResult result = new API().execProgram(args);

        if (!result.hasSourcemap() || result.getSourcemap().isEmpty()) {
            Assume.assumeNoException(
                    "runtime did not populate sourcemap (kcl-api may not yet support source maps)",
                    null);
        }
        Assert.assertTrue(
                "sourcemap_output: result.sourcemap must contain Source Map version key",
                result.getSourcemap().contains("\"version\""));
    }
}

File: kotlin/src/test/kotlin/com/kcl/ExecProgramFormatTest.kt
package com.kcl

import com.kcl.api.API
import com.kcl.api.Spec.ExecProgramArgs
import com.kcl.api.Spec.ExecProgramResult
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assumptions.assumeTrue
import org.junit.jupiter.api.Test

/**
 * Tests for the source-map fields introduced alongside
 * kcl-lang/kcl#1546. Combines:
 *   1. Pure protobuf round-trip — verifies the regenerated Spec.java
 *      encodes / decodes the new fields without touching the runtime.
 *   2. End-to-end execution — runs the bundled schema.k through
 *      API().execProgram with format="json" and sourcemap_output set,
 *      asserting on the actual runtime output.
 *
 * Covers:
 *   - ExecProgramArgs.format           (field 20)
 *   - ExecProgramArgs.error_format     (field 19, pre-existing sanity check)
 *   - ExecProgramArgs.sourcemap_output (field 22)
 *   - ExecProgramResult.sourcemap      (field 5)
 */
class ExecProgramFormatTest {

    companion object {
        private const val TEST_FILE = "./src/test_data/schema.k"
    }

    @Test
    fun testExecProgramArgsFormatRoundTrip() {
        val args = ExecProgramArgs.newBuilder()
            .setFormat("json")
            .setErrorFormat("sarif")
            .setSourcemapOutput("/tmp/out.js.map")
            .build()

        val decoded = ExecProgramArgs.parseFrom(args.toByteString())

        assertEquals("json", decoded.format)
        assertEquals("sarif", decoded.errorFormat)
        assertTrue(decoded.hasSourcemapOutput(), "sourcemap_output presence")
        assertEquals("/tmp/out.js.map", decoded.sourcemapOutput)
    }

    @Test
    fun testExecProgramResultSourcemapRoundTrip() {
        val result = ExecProgramResult.newBuilder()
            .setJsonResult("{\"a\": 1}")
            .setYamlResult("a: 1")
            .setSourcemap("{\"version\":3,\"sources\":[]}")
            .build()

        val decoded = ExecProgramResult.parseFrom(result.toByteString())

        assertEquals("{\"a\": 1}", decoded.jsonResult)
        assertEquals("a: 1", decoded.yamlResult)
        assertTrue(decoded.hasSourcemap(), "sourcemap presence")
        assertEquals("{\"version\":3,\"sources\":[]}", decoded.sourcemap)
    }

    // End-to-end: format=json must populate only json_result.
    @Test
    fun testExecProgramFormatJson() {
        val args = ExecProgramArgs.newBuilder()
            .addKFilenameList(TEST_FILE)
            .setFormat("json")
            .build()
        val result = API().execProgram(args)

        assertFalse(result.jsonResult.isEmpty(), "format=json: json_result must be populated")
        assertEquals("", result.yamlResult, "format=json: yaml_result must be empty")
        assertTrue(result.jsonResult.contains("replicas"), "format=json: json_result must contain replicas key")
    }

    // End-to-end: sourcemap_output must populate result.sourcemap. Older
    // runtimes that don't yet emit source maps are skipped via assumeTrue
    // so this test doesn't fail under a stale kcl-api.
    @Test
    fun testExecProgramSourcemapOutput() {
        val args = ExecProgramArgs.newBuilder()
            .addKFilenameList(TEST_FILE)
            .setSourcemapOutput("/tmp/out.js.map")
            .build()
        val result = API().execProgram(args)

        assumeTrue(
            result.hasSourcemap() && result.sourcemap.isNotEmpty(),
            "runtime did not populate sourcemap (kcl-api may not yet support source maps)",
        )
        assertTrue(
            result.sourcemap.contains("\"version\""),
            "sourcemap_output: result.sourcemap must contain Source Map version key",
        )
    }
}